src/app/shared/doc-viewer/doc-viewer.ts

```diff
@@ -15,18 +16,34 @@ import {
   SecurityContext,
   ViewContainerRef,
 } from '@angular/core';
-import {Subscription} from 'rxjs';
-import {take} from 'rxjs/operators';
+import {Observable, Subscription} from 'rxjs';
+import {shareReplay, take, tap} from 'rxjs/operators';
 import {ExampleViewer} from '../example-viewer/example-viewer';
 import {HeaderLink} from './header-link';
 
+@Injectable({providedIn: 'root'})
+class DocFetcher {
+  private _cache: Record<string, Observable<string>> = {};
+
+  constructor(private _http: HttpClient) {}
+
+  fetchDocument(url: string): Observable<string> {
+    if (this._cache[url]) {
+      return this._cache[url];
+    }
+
+    const stream = this._http.get(url, {responseType: 'text'}).pipe(shareReplay(1));
+    return stream.pipe(tap(() => this._cache[url] = stream));
```

Comment:
My expertise with RxJS is limited, but is there are a reason to cache the stream in the `tap`? Is this to avoid caching cases where an initial XHR request failed?

Reply:
Yeah, I was trying to avoid caching a failed request.